third_party/amd/backend/compiler.py

```
@@ -70,6 +70,9 @@ def __post_init__(self):
         # Ignore user-defined warp size for gfx9
         warp_size = 32 if 'gfx10' in self.arch or 'gfx11' in self.arch or 'gfx12' in self.arch else 64
         object.__setattr__(self, 'warp_size', warp_size)
+        # Cap warp count to 16 on devices with warp size 64 (1024 threads/block)
```

Theoretically the max threads is architecture specific so we should query it with `gfx*` target. [In reality I think 1024 is the value](https://github.com/llvm/llvm-project/blob/eac734aab6b62518a0cd7aa3b20a990fef360b2c/llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp#L1036). We should error out instead of silently rewriting user provided values. And there is no need to gate on warp size 64; you can deduce the max num warps by with 1024 / warp_size.

---

Changed to an assert. I guess it should be the user's responsibility to make sure they're checking the device warp size, but I would prefer if there was some way to override this (i.e. capping the warp count or decreasing the warp size), even if non-optimal. Still, having a descriptive assert is better than the status quo.